### ISSUE TITLE
Setting SO_LINGER completely inconsistent

### ISSUE THREAD

**Issue by etehtsea** [jruby/jruby#4040]
MRI works / JRuby unsupported:
`s.setsockopt(Socket::Option.linger(true, 0))`

----------------------------------------------------

MRI works / JRuby accepted, but does nothing (broken):

```
require 'socket'; s = Socket.tcp('localhost', 6379); s.setsockopt(Socket::SOL_SOCKET, Socket::SO_LINGER, [1,0].pack('ii')); s.close
```

MRI:

```
17358	7700.821842	127.0.0.1	127.0.0.1	TCP	44	62123 → 6379 [RST, ACK] Seq=1 Ack=1 Win=408288 Len=0
```

JRuby:

```
17291	7541.692073	127.0.0.1	127.0.0.1	TCP	56	62096 → 6379 [FIN, ACK] Seq=1 Ack=1 Win=408288 Len=0 TSval=457027447 TSecr=457027443
17292	7541.692126	127.0.0.1	127.0.0.1	TCP	56	6379 → 62096 [ACK] Seq=1 Ack=2 Win=408288 Len=0 TSval=457027447 TSecr=457027447
17293	7541.692133	127.0.0.1	127.0.0.1	TCP	56	[TCP Dup ACK 17289#1] 62096 → 6379 [ACK] Seq=2 Ack=1 Win=408288 Len=0 TSval=457027447 TSecr=457027447
17294	7541.692164	127.0.0.1	127.0.0.1	TCP	56	6379 → 62096 [FIN, ACK] Seq=1 Ack=2 Win=408288 Len=0 TSval=457027447 TSecr=457027447
17295	7541.692252	127.0.0.1	127.0.0.1	TCP	56	62096 → 6379 [ACK] Seq=2 Ack=2 Win=408288 Len=0 TSval=457027447 TSecr=457027447
```

------------------------------------------------

MRI unsupported / JRuby broken:

MRI:

```
[7] pry(main)> require 'socket'; s = Socket.tcp('localhost', 6379); s.setsockopt(Socket::SOL_SOCKET, Socket::SO_LINGER, 0); s.close
Errno::EINVAL: Invalid argument - setsockopt(2)
from (pry):7:in `setsockopt'
```

JRuby same as previous.

--------------------------------------------------

MRI unsupported / JRuby actually works:

```
require 'socket'; s = Socket.tcp('localhost', 6379); s.setsockopt(Socket::SOL_SOCKET, :LINGER, 0); s.close
```

MRI same as previous.

JRuby:

```
require 'socket'; s = Socket.tcp('localhost', 6379); s.setsockopt(Socket::SOL_SOCKET, :LINGER, 0); s.close
=> nil
```

```
17553	8085.477203	127.0.0.1	127.0.0.1	TCP	44	62165 → 6379 [RST, ACK] Seq=1 Ack=1 Win=408288 Len=0
```

**Comment by headius** [comment]
@etehtsea This is fixed now, right?

**Comment by etehtsea** [comment]
@headius nope. Even in https://github.com/jruby/jruby/pull/4093

**Comment by etehtsea** [comment]
@headius I've prepared PR #4119 that fixes this issue.

**Comment by enebo** [comment]
@etehtsea we will try and get the branch you and @headius have been working on landed early in 9.1.4.0 but it will miss 9.1.3.0
